leader-election-impl/src/main/java/com/palantir/leader/proxy/LeadershipStateManager.java

```diff
@@ -0,0 +1,185 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.leader.proxy;
+
+import com.palantir.common.concurrent.CoalescingSupplier;
+import com.palantir.common.concurrent.PTExecutors;
+import com.palantir.leader.LeaderElectionService.LeadershipToken;
+import com.palantir.leader.NotCurrentLeaderException;
+import com.palantir.logsafe.SafeArg;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LeadershipStateManager<T> {
+    private static final Logger log = LoggerFactory.getLogger(LeadershipStateManager.class);
+
+    private final CoalescingSupplier<LeadershipToken> leadershipTokenCoalescingSupplier;
+    private final LeadershipCoordinator leadershipCoordinator;
+    private final AtomicReference<T> delegateRef;
+    private final AtomicReference<LeadershipToken> maybeValidLeadershipTokenRef;
+    private final Supplier<T> delegateSupplier;
+    private final ExecutorService delegateClosingExecutor;
+
+    private volatile boolean isClosed;
+
+    public LeadershipStateManager(LeadershipCoordinator leadershipCoordinator, Supplier<T> delegateSupplier) {
+        this.leadershipTokenCoalescingSupplier = new CoalescingSupplier<>(this::getOrUpdateLeadershipToken);
+        this.leadershipCoordinator = leadershipCoordinator;
+        this.delegateSupplier = delegateSupplier;
+        this.delegateRef = new AtomicReference<>();
+        this.maybeValidLeadershipTokenRef = new AtomicReference<>();
+        this.delegateClosingExecutor = PTExecutors.newSingleThreadExecutor();
+        this.isClosed = false;
+    }
+
+    LeadershipState<T> getLeadershipState() {
+        return ImmutableLeadershipState.<T>builder()
+                .leadershipToken(leadershipTokenCoalescingSupplier.get())
+                .delegate(delegateRef.get())
+                .build();
+    }
+
+    void close() {
+        isClosed = true;
+        delegateClosingExecutor.shutdownNow();
+        clearDelegate();
+    }
+
+    /**
+     * Checks the local ref of leadership token. We need to refresh the delegateRef in case the token has
+     * changed.
+     * This code can be accessed by multiple threads. In order to save all of the requests from locking-unlocking
+     * `this` while trying to update maybeValidLeadershipTokenRef, the invocations are batched using CoalescingSupplier.
+     */
+    @GuardedBy("leadershipTokenCoalescingSupplier")
```

Comment:
Hmm, not exactly: this has a specific meaning which Errorprone might not be too happy about

Reply:
this would mean we get a lock on `leadershipTokenCoalescingSupplier `